File: mobile/ios/packages/town-crier-data/Sources/Repositories/APIUserProfileRepository.swift
import Foundation
import TownCrierDomain

/// Manages the user's server-side profile via the Town Crier API.
public final class APIUserProfileRepository: UserProfileRepository, Sendable {
  private let apiClient: URLSessionAPIClient

  public init(apiClient: URLSessionAPIClient) {
    self.apiClient = apiClient
  }

  public func create() async throws -> ServerProfile {
    do {
      let dto: ServerProfileDTO = try await apiClient.request(.post("/v1/me"))
      return dto.toDomain()
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
  }

  public func fetch() async throws -> ServerProfile? {
    do {
      let dto: ServerProfileDTO = try await apiClient.request(.get("/v1/me"))
      return dto.toDomain()
    } catch APIError.notFound {
      return nil
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
  }

  public func update(
    pushEnabled: Bool,
    digestDay: DayOfWeek,
    emailDigestEnabled: Bool
  ) async throws -> ServerProfile {
    let body = UpdateProfileRequest(
      pushEnabled: pushEnabled,
      digestDay: digestDay,
      emailDigestEnabled: emailDigestEnabled
    )
    do {
      let dto: ServerProfileDTO = try await apiClient.request(
        .patch("/v1/me", body: body)
      )
      return dto.toDomain()
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
  }

  public func delete() async throws {
    do {
      let _: EmptyResponse = try await apiClient.request(.delete("/v1/me"))
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
  }
}

// MARK: - DTOs

struct ServerProfileDTO: Decodable, Sendable {
  let userId: String
  let tier: String
  let pushEnabled: Bool
  let digestDay: String
  let emailDigestEnabled: Bool

  enum CodingKeys: String, CodingKey {
    case userId
    case tier
    case pushEnabled
    case digestDay
    case emailDigestEnabled
  }

  init(from decoder: any Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    userId = try container.decode(String.self, forKey: .userId)
    tier = try container.decode(String.self, forKey: .tier)
    pushEnabled = try container.decode(Bool.self, forKey: .pushEnabled)
    digestDay = try container.decodeIfPresent(String.self, forKey: .digestDay) ?? "Monday"
    emailDigestEnabled =
      try container.decodeIfPresent(Bool.self, forKey: .emailDigestEnabled)
      ?? true
  }

  func toDomain() -> ServerProfile {
    ServerProfile(
      userId: userId,
      tier: SubscriptionTier(rawValue: tier.lowercased()) ?? .free,
      pushEnabled: pushEnabled,
      digestDay: DayOfWeek(rawValue: digestDay) ?? .monday,
      emailDigestEnabled: emailDigestEnabled
    )
  }
}

struct UpdateProfileRequest: Encodable, Sendable {
  let pushEnabled: Bool
  let digestDay: DayOfWeek
  let emailDigestEnabled: Bool

  enum CodingKeys: String, CodingKey {
    case pushEnabled
    case digestDay
    case emailDigestEnabled
  }
}

File: mobile/android/data/src/main/kotlin/uk/towncrierapp/data/profile/ApiUserProfileRepository.kt
package uk.towncrierapp.data.profile

import kotlinx.serialization.Serializable
import uk.towncrierapp.data.api.ApiClient
import uk.towncrierapp.data.api.ApiEndpoint
import uk.towncrierapp.domain.profile.ServerProfile
import uk.towncrierapp.domain.profile.UserProfileRepository
import uk.towncrierapp.domain.subscriptions.SubscriptionTier

/** `UserProfileRepository` over the Town Crier API — `POST /v1/me` (ensure-profile) only; fetch/update/delete/export land with #778. */
public class ApiUserProfileRepository(
    private val apiClient: ApiClient,
) : UserProfileRepository {
    override suspend fun ensureProfile(): ServerProfile =
        apiClient.request(ApiEndpoint.post("/v1/me"), ServerProfileDto.serializer()).toDomain()
}

@Serializable
internal data class ServerProfileDto(
    val userId: String,
    val pushEnabled: Boolean,
    val tier: String,
)

internal fun ServerProfileDto.toDomain(): ServerProfile =
    ServerProfile(
        userId = userId,
        pushEnabled = pushEnabled,
        // Falls back to Free on an unrecognised wire value rather than crashing —
        // the server is authoritative and may add tiers this build doesn't know yet.
        tier = SubscriptionTier.fromWireValue(tier) ?: SubscriptionTier.FREE,
    )
